### DISCUSSION TITLE
How to match keys by partial name when location isn't known?

### DISCUSSION THREAD

**Discussion by steve-nay-sage** [mikefarah/yq#955] [Q&A] [answered]
I'm working with some OpenAPI definitions that have `x-` properties at various locations throughout the files. For example, there could be an `x-mappedTo` property as a top-level property, and/or there could be other `x-whatever` properties in other places in the file at other levels of nesting. I need to remove those properties/entries. So I need to find all entries whose key value starts with x- regardless of where they are in the file. 

- There is no way to know the property hierarchy where these will be found.
- There is no way to know the full property name (key); it's just going to start with `x-`.

I've attached the top part of an OpenAPI file that contains a few of these properties. It's YAML, even though the file name is .txt.

I hack around at programming and write API documentation, but I haven't been able to figure out how to do this using yq. I've tried some of the examples to get my feet wet, and some work and some don't. One that doesn't work is 

```
yq eval 'with_entries(select(.value | has("b")))' sample.yml
```

Just returns an empty set when my sample.yaml file definitely contains values that have "b" in them. So I would greatly appreciate any help or insights that anyone can shar
[x-test.txt](https://github.com/mikefarah/yq/files/7297535/x-test.txt)
e.

**Comment by mikefarah** [comment, marked as answer]
You're on the right track with that, but first you have to tell it to recurse though the whole structure. So you will need to do somthing like:

```
yq e '.. | select(tag == "!!map") | with_entries(select(.key | match("x-.*")))' examples/data1.yaml
```

Explanation:
1. `..` tells yq to recursively match all the nodes
2. `select(tag == "!!map")` we only select maps (as arrays wont have `x-` keys anyway
3. `with_entries(select(.key | match("x-.*")))` we filter out the maps to only include the keys that start with x-

**Reply by mikefarah** [reply]
Forgot to put the output of that:

```
x-mappedTo: vendor
x-mappedTo: RECORDNO
x-mappedTo: VENDORID
x-mutable: false
x-mappedTo: NAME
```

**Reply by steve-nay-sage** [reply]
Thanks, @mikefarah. I'm sorry to be dense, but I'm not sure how to use this output to effectively delete these properties once they are located in the file. That is, my goal is to have the original file without the x- properties.

I'm just not getting the hang of the yq syntax, and I'm not finding the jq docs to be much help either. The jq "match" filter doesn't seem to be at all like how you've used "match" in this expression. I've spent quite a while on this today and I've only succeeded in making myself feel stupid and a bit frustrated. Is there some metaphor that underlies yq (and jq) that would help me understand the pretty terse format? I've use xpath in the past and haven't stumbled like this.

Thanks again. -Steve

**Reply by mikefarah** [reply]
Yeah it took me quite a while to wrap my head around it too - I've only really learnt how `jq` works by trying to re-implement it - and then trying to help people with their yaml tasks - and of course I'm still learning. I also came from an xpath background, and this is quite different. This is more akin to using bash, piping data through lots of little command line utils to get what you want.

I tried to document the underlying principals here - would be good to get your feedback on it: https://mikefarah.gitbook.io/yq/operators

Here's how you can remove the matching keys:

```
yq e '(.. | select(tag=="!!map")) |= with_entries(select(.key | test("x-.*") | not))' examples/data1.yaml
```

Explanation:

1. The `|=` is a relative update operator, it takes the LHS and updates each node relative to itself (https://mikefarah.gitbook.io/yq/operators/assign-update#double-elements-in-an-array)
2. On the LHS - we want to update all the maps in the yaml object
3. The RHS is now going to run for each maching LHS, that is every map in the yaml. We can use with_entries (https://mikefarah.gitbook.io/yq/operators/assign-update#double-elements-in-an-array) to filter out the keys you don't want.

**Reply by steve-nay-sage** [reply]
Thanks for the quick response. I'm encouraged because I almost had it right--I had the "| not" in the right place, but hadn't replaced match with test. 
It seems odd to me that this command is so dramatically different from deleting a node when you know it's location and name, which would be just 

```
yq e '(del(.a.b.c)' file.yaml
```
The command you gave me doesn't even use the delete operator, so it's easy to see how someone (me) could head off in the wrong direction with this one. 

Have you considered an operation similar to `find` in the shell? Maybe something like this:

```
yq e '(del(find . -key "x-*"))' file.yaml
```

I see some parallels there, but I could be way off.

I'll put the command that you gave me to work, and then hopefully I'll be able to come back to this and give you some feedback on the documentation.

**Reply by steve-nay-sage** [reply]
@mikefarah It's been a while since you helped me with this. I've upgraded yq to v4.25.3, and I can't get this yq script to work quite right. I've puzzled over it and tried lots of variations, but still can't fix it. The goal is to find all keys that start with x- and remove them and their child elements. This script finds the x- keys if they are at the top level of the hierarchy, but it does not find them anywhere else. For example, in this file:

```
candy: chocolate
x-a:
  b:
    c: some value
b:
  foo: bar
  x-b:
    some: thing
    x-z: a simple x- key-value pair
    x-c:
      one: more
```
It matches and removes x-a, but not x-b or x-z or x-c. 

I would really appreciate your help with this, as I've spent an embarrassing amount of time on it without success.

**Reply by mikefarah** [reply]
Hi Steve, not a problem:

TLDR:

```bash
yq 'del(.. | select(key | test("^x-")))' file.yaml
```

result:
```yaml
candy: chocolate
b:
  foo: bar
```

Explanation:
We need to select all the nodes that have a key beginning with "x-" and delete them.
- `..` recursively matches all nodes
- Piping that through to `select(key | test("^x-"))` will filter that to all nodes with a key that begins with "x-"
- del( ) will delete those nodes

**Reply by steve-nay-sage** [reply]
Wow, thank you for that amazingly fast answer. Your answer looks so elegant, too! I was trying to make this much harder that it is. I'll try to put this to good use tomorrow.

**Reply by mikefarah** [reply]
You caught me just as I was checking to see if there was any yq action :)

**Reply by steve-nay-sage** [reply]
@mikefarah I didn't get far today. In this case I need to get values rather than delete something, so I removed the del() part of the request. That works fine on the little test file that I sent you, but when I apply it to something more like the real data that I'm working with it returns this error:

```
> yq '.. | select(key | test("^x-"))' file.yaml

Error: cannot match with !!int, can only match strings.
```

I don't see anything that looks like an int in the file, but I'm sure that I'm not reading it right. An example of the real data that I'm working with is this, and I want to be able to find x-permissions and work with what's under it:

```
paths:
  /objects/class:
    get:
      summary: List classes
      description: >
        Returns up to 100 object references from the collection with a key, ID,
        and link for each object. 
      x-permissions:
        subscription: Company
        userPermissions:
          - userType: Business user with admin privileges
            permissions: List, View, Add, Edit, Delete
          - userType: Employee
            permissions: List, View
          - userType: Approval
            permissions: List, View
      tags:
        - Classes
```

That pattern could occur several times in each file that I need to process. To work with each one separately, do I need to use a bash `for` loop? Of course, before getting to that part, I first need to find x-permissions in the file. This is what I have, but is it the best way?

`yq '.. | select(has("x-permissions")) | .x-permissions' file.yaml`

**Reply by mikefarah** [reply]
Oh my bad - forgot about arrays (which have `int` keys:

```
yq '.. | select(( key | tag == "!!str" ) and  (key | test("^x-")))' file.yaml
```

Yeah a bash for loop or find with xargs 👍🏼

**Reply by steve-nay-sage** [reply]
Thank you @mikefarah -- I have my script working, and I wouldn't have been able to do it without your help.
